pkg/vgmanager/lvm.go

```diff
@@ -23,33 +23,206 @@ import (
+	} `json:"report"`
+}
+
+// pvsOutput represents the output of the `pvs --reportformat json` command
+type pvsOutput struct {
+	Report []struct {
+		Pv []struct {
+			Name   string `json:"pv_name"`
+			VgName string `json:"vg_name"`
+		} `json:"pv"`
+	} `json:"report"`
+}
+
 // VolumeGroup represents a volume group of linux lvm.
 type VolumeGroup struct {
+	// Name is the name of the volume group
 	Name string `json:"vg_name"`
+
+	// PVs is the list of physical volumes associated with the volume group
+	PVs []string `json:"pvs"`
 }
 
-// ListVolumeGroups lists all volume groups.
-func ListVolumeGroups(exec internal.Executor) ([]VolumeGroup, error) {
-	// volumeGroupMap type is used for marshalling the `vgs --reportformt json` command output
-	// Sample output:
-	// `{"report": [ {"vg": [{"vg_name":"fedora", "pv_count":"1", "lv_count":"3", "snap_count":"0", "vg_attr":"wz--n-", "vg_size":"<475.94g", "vg_free":"0 "}]}]`
-	volumeGroupMap := map[string][]map[string][]VolumeGroup{}
-	arg := []string{
+// Create creates a new volume group
+func (vg VolumeGroup) Create(exec internal.Executor, pvs []string) error {
+	if vg.Name == "" {
+		return fmt.Errorf("failed to create volume group. Volume group name is empty")
+	}
+
+	if len(pvs) == 0 {
+		return fmt.Errorf("failed to create volume group. Physical volume list is empty")
+	}
+
+	args := []string{vg.Name}
+	args = append(args, pvs...)
+
+	_, err := exec.ExecuteCommandWithOutputAsHost(vgCreateCmd, args...)
+	if err != nil {
+		return fmt.Errorf("failed to create volume group %q. %v", vg.Name, err)
+	}
+
+	return nil
+}
+
+// Extend extends the volume group only if new physical volumes are available
+func (vg VolumeGroup) Extend(exec internal.Executor, pvs []string) error {
+	if vg.Name == "" {
+		return fmt.Errorf("failed to extend volume group. Volume group name is empty")
+	}
+
+	if len(pvs) == 0 {
+		return fmt.Errorf("failed to extend volume group. Physical volume list is empty")
+	}
+
+	args := []string{vg.Name}
+	args = append(args, pvs...)
+
+	_, err := exec.ExecuteCommandWithOutputAsHost(vgExtendCmd, args...)
+	if err != nil {
+		return fmt.Errorf("failed to extend volume group %q. %v", vg.Name, err)
+	}
+
+	return nil
+}
+
+// Delete deletes a volume group and the physical volumes associated with it
+func (vg VolumeGroup) Delete(exec internal.Executor) error {
+	// Remove Volume Group
+	vgArgs := []string{vg.Name}
+	_, err := exec.ExecuteCommandWithOutputAsHost(vgRemoveCmd, vgArgs...)
+	if err != nil {
+		return fmt.Errorf("failed to remove volume group %q. %v", vg.Name, err)
+	}
+
+	// Remove physical volumes
+	_, err = exec.ExecuteCommandWithOutputAsHost(pvRemoveCmd, vg.PVs...)
+	if err != nil {
+		return fmt.Errorf("failed to remove physical volumes for the volume group %q. %v", vg.Name, err)
+	}
+	return nil
+}
+
+// GetVolumeGroup returns a volume group along with the associated physical volumes
+func GetVolumeGroup(exec internal.Executor, name string) (*VolumeGroup, error) {
```

Comment:
- Is this function a standby for now as I don't see it's usage?
- I can see some functions as this only with tests but this function seems to be a duplicate or specific case of `ListVolumeGroups`?

Reply:
Its not used anywhere now.  Since we might move to  one CR for each VG now, we can replace `ListVolumeGroup` with `GetVolumeGroup`.   `GetVolumeGroup` checks if volume groups is not available.